File: ga/gpt3_5_result_2024_6/result/1968F_cpp_gpt/22_1968F.cpp
use std::cmp::Ordering;
use std::io::{self, BufRead};

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines().map(|line| line.unwrap());

    let t: usize = lines.next().unwrap().parse().unwrap();
    for _ in 0..t {
        let line: String = lines.next().unwrap();
        let mut parts = line.split_whitespace();
        let n: usize = parts.next().unwrap().parse().unwrap();
        let q: usize = parts.next().unwrap().parse().unwrap();

        let mut arr: Vec<i32> = Vec::new();
        let line: String = lines.next().unwrap();
        arr.extend(line.split_whitespace().map(|x| x.parse().unwrap()));

        let mut x: Vec<(i32, usize)> = vec![(0, 0); n+1];
        x[0] = (0, 0);
        for i in 1..=n {
            x[i] = (x[i-1].0 ^ arr[i-1], i);
        }

        let mut x2: Vec<i32> = Vec::new();
        x2.extend(x.iter().map(|&a| a.0));
        x.sort();

        for _ in 0..q {
            let line: String = lines.next().unwrap();
            let mut parts = line.split_whitespace();
            let l: usize = parts.next().unwrap().parse().unwrap();
            let r: usize = parts.next().unwrap().parse().unwrap();

            let c = x2[r] ^ x2[l-1];
            if c == 0 {
                println!("YES");
                continue;
            }
            match x.binary_search_by(|a| a.cmp(&(c ^ x2[l-1]))) {
                Ok(i1) => {
                    if i1 == n+1 || x[i1].1 > r || x[i1].0 != (c ^ x2[l-1]) {
                        println!("NO");
                        continue;
                    }
                    match x[i1+1..].binary_search_by(|a| a.1.cmp(&r)) {
                        Ok(i2) => {
                            if x[i1+i2+1].0 == x2[l-1] {
                                println!("YES");
                            } else {
                                println!("NO");
                            }
                        }
                        _ => println!("NO")
                    }
                }
                _ => println!("NO")
            }
        }
    }
}

File: ga/glm4_result_2024_6/result/1968F_cpp_glm/22_1968F.rs
use std::io::{self, BufRead};
use std::collections::BinaryHeap;

#[derive(Eq, PartialEq)]
struct Query {
    l: usize,
    r: usize,
    i: usize,
}

impl Query {
    fn new(l: usize, r: usize, i: usize) -> Self {
        Query { l, r, i }
    }
}

impl Ord for Query {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.r.cmp(&other.r)
    }
}

impl PartialOrd for Query {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl std::fmt::Debug for Query {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Query")
            .field("l", &self.l)
            .field("r", &self.r)
            .field("i", &self.i)
            .finish()
    }
}

fn solve() {
    let stdin = io::stdin();
    let lines = stdin.lock().lines();
    let mut lines = lines.map(|line| line.unwrap());

    let (n, q) = lines.next().unwrap().split_whitespace().map(|s| s.parse().unwrap()).collect_tuple().unwrap();
    let arr = lines.next().unwrap().split_whitespace().map(|s| s.parse().unwrap()).collect::<Vec<_>>();

    let mut x = vec![0; n + 1];
    for i in 1..=n {
        x[i] = x[i - 1] ^ arr[i - 1];
    }
    let x2 = x.clone();

    let mut heap = BinaryHeap::new();
    for (i, &r) in x.iter().enumerate() {
        heap.push(Query::new(i, i, i));
    }

    while let Some(Query { l, r, i }) = heap.pop() {
        if l > r {
            continue;
        }
        if x[l] == x2[r] {
            println!("YES");
        } else {
            let l1 = x2[l].checked_shr(1).map(|v| *heap.get_binary_heap().binary_search_by(|&Query { l: l1, r: r1, i: _ }| l1.cmp(&v)).unwrap());
            let r1 = x2[r].checked_shr(1).map(|v| *heap.get_binary_heap().binary_search_by(|&Query { l: l1, r: r1, i: _ }| r1.cmp(&v)).unwrap());
            if l1.is_some() && r1.is_some() {
                println!("YES");
            } else {
                println!("NO");
            }
        }
    }
}

fn main() {
    let stdin = io::stdin();
    let lines = stdin.lock().lines();
    let mut lines = lines.map(|line| line.unwrap());

    let t = lines.next().unwrap().parse().unwrap();
    for _ in 0..t {
        solve();
    }
}